### PR TITLE
Change search by in autocomplete widget

### PR DESCRIPTION
**Pull request by szok**
- Add polymorphic descendants search fields names

### CODE REVIEW

**Comment by szok** [comment]
zrzut ekranu 2015-11-27 o 14 44 21

**Comment by mkurek** [comment]
Add new lines (`<br>`) in tooltip

**Comment by szok** [comment]
zrzut ekranu 2015-11-27 o 15 36 31

**Comment by szok** [comment]
zrzut ekranu 2015-11-27 o 15 39 17

**Comment by mkurek** [comment]
Is sth like this possible:

```
Back office asset:
  - SN
  - barcode
  - hostname
Data center asset:
  - SN
  - barcode
  - hostname
Licence:
  - ...
```

And just for the record, we should implement here sth like `limit_choices_to` - Licence (using BaseObject) could be assigned only to DC Asset, BO Asset, Virtual Server etc (but definately not another Licence, Support and ServiceEnv)

**Comment by szok** [comment]
screen shot 2015-12-02 at 13 50 48

File: src/ralph/lib/mixins/tests/test_fields.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+from django.test import TestCase
+
+from ralph.assets.models.base import BaseObject
+from ralph.back_office.models import BackOfficeAsset
+from ralph.data_center.models import DataCenterAsset
+from ralph.tests.models import BaseObjectForeignKeyModel
+
+
+class BaseObjectForeignKeyTestCase(TestCase):
+    def test_limit_choics(self):
```
Comment (mkurek) [inline src/ralph/lib/mixins/tests/test_fields.py]: choices

File: src/ralph/admin/widgets.py [hunk 2]
```diff
@@ -183,12 +190,43 @@ def render(self, name, value, attrs=None):
...
-                    field = get_field_by_relation_path(self.rel_to, field_name)
-                    searched_fields.append(str(field.verbose_name))
-                except FieldDoesNotExist:
-                    pass
+            searched_fields_tooltip = defaultdict(list)
+            polymorphic_descendants = getattr(
+                self.rel_to, '_polymorphic_descendants', []
+            )
+
+            if polymorphic_descendants:
```
Comment (mkurek) [inline src/ralph/admin/widgets.py:198]: please add some comments what is going on here

File: src/ralph/admin/autocomplete.py [hunk 3]
```diff
@@ -140,14 +86,70 @@ def get_queryset(self, user):
...
+            )
+        else:
+            queryset = model.objects
+        return queryset.filter(
+            *query_filters
+        )[:self.limit].values_list('pk', flat=True)
+
+    def get_queryset(self, user):
+        search_fields = ralph_site._registry[self.model].search_fields
+        queryset = self.model._default_manager.all()
```
Comment (mkurek) [inline src/ralph/admin/autocomplete.py]: could you add ability to easily overwrite queryset? I'll need this in https://github.com/allegro/ralph/pull/2033/files#diff-c920c22878885266492749084f889519R63

File: src/ralph/admin/autocomplete_urls.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# -*- coding: utf-8 -*-
+from django.conf.urls import url
+
+from ralph.admin.autocomplete import AutocompleteList
+
+urlpatterns = [
+    url(
+        r'^autocomplete/(?P<app>\w+)/(?P<model>\w+)/(?P<field>\w+)',
```
Comment (mkurek) [inline src/ralph/admin/autocomplete_urls.py]: wouldn't it be better to use following url:

```
^(?P<app>\w+)/(?P<model>\w+)/(?P<field>\w+)/autocomplete$
```

Then `^(?P<app>\w+)/(?P<model>\w+)/(?P<field>\w+)` could be used to display list of available items after you click on magnifier icon next to autocomplete.

@ar4s what do you think?

File: src/ralph/admin/autocomplete.py [hunk 5]
```diff
@@ -140,14 +86,70 @@ def get_queryset(self, user):
...
         ]
         return my_urls + urls
+
+
+class AutocompleteList(SuggestView):
+    limit = 10
+    model = None
+
+    def dispatch(self, request, *args, **kwargs):
+        model = get_model(kwargs['app'], kwargs['model'])
```
Comment (ar4s) [inline src/ralph/admin/autocomplete.py]: please handle `LookupError` - it raising when model doesn't not found

File: src/ralph/admin/autocomplete_urls.py (hunk 4 above)
Comment (ar4s) [inline src/ralph/admin/autocomplete_urls.py]: +1

File: src/ralph/admin/autocomplete_urls.py [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+        AutocompleteList.as_view(),
```
Comment (ar4s) [inline src/ralph/admin/autocomplete_urls.py]: please wrap this by `admin_view` from `ralph.admin.sites.ralph_site`

File: src/ralph/admin/widgets.py [hunk 7]
```diff
@@ -159,7 +162,11 @@ def render(self, name, value, attrs=None):
         )
         widget_options = {
             'data-suggest-url': reverse(
-                'admin:{}_{}_autocomplete_suggest'.format(*model_options)
+                'autocomplete-list', kwargs={
+                    'app': self.field.rel.related_model._meta.app_label,
+                    'model': self.field.rel.related_model.__name__,
```
Comment (mkurek) [inline src/ralph/admin/widgets.py]: are you sure that `__name__` here is ok? what about `_meta.model_name`?

File: src/ralph/lib/mixins/fields.py [hunk 8]
```diff
@@ -30,3 +32,20 @@ class NullableGenericIPAddressField(
     models.GenericIPAddressField
 ):
     pass
+
+
+class BaseObjectForeignKey(models.ForeignKey):
```
Comment (mkurek) [inline src/ralph/lib/mixins/fields.py:37]: add docstring

**Comment by mkurek** [comment]
:ok:

File: src/ralph/admin/autocomplete.py [hunk 9]
```diff
@@ -144,14 +89,89 @@ def get_queryset(self, user):
...
+
+class AutocompleteList(SuggestView):
+    limit = 10
+    model = None
+
+    def dispatch(self, request, *args, **kwargs):
+        try:
+            model = get_model(kwargs['app'], kwargs['model'])
+        except LookupError:
+            return HttpResponseBadRequest('Model not found')
```
Comment (ar4s) [inline src/ralph/admin/autocomplete.py]: App or model not found

**Comment by ar4s** [comment]
:ok:

**Pull request merged**